src/diamonds/facets/SnapshotsLibrary.sol

```diff
@@ -102,60 +104,102 @@ library SnapshotsLibrary {
-            );
-
+            ethBlocksSinceLastSnapshot = block.number - previousSnapshot.ethHeight;
+            
+            if (cs.epoch > 1) {
+
+                require(
+                    !previousSnapshot.saved || ethBlocksSinceLastSnapshot >= ss.minEthSnapshotSize,
+                    "snapshot heights too close in Ethereum"
+                );
+
+                require(
+                    !previousSnapshot.saved || madHeight - previousSnapshot.madHeight >= ss.minMadSnapshotSize,
+                    "snapshot heights too close in MadNet"
+                );
+            }
         }
 
+        // Lock rewards
         ParticipantsLibrary.ParticipantsStorage storage ps = ParticipantsLibrary.participantsStorage();
-        StakingLibrary.StakingStorage storage stakingS = StakingLibrary.stakingStorage();
-
-        for (uint idx=0; idx<ps.validators.length; idx++) {
-            if (msg.sender==ps.validators[idx]) {
-                StakingLibrary.lockRewardFor(ps.validators[idx], stakingS.rewardAmount + stakingS.rewardBonus, cs.epoch+2);
-            } else {
-                StakingLibrary.lockRewardFor(ps.validators[idx], stakingS.rewardAmount, cs.epoch+2);
+        {
+            StakingLibrary.StakingStorage storage stakingS = StakingLibrary.stakingStorage();
+            for (uint idx=0; idx<ps.validators.length; idx++) {
+                if (msg.sender==ps.validators[idx]) {
+                    StakingLibrary.lockRewardFor(ps.validators[idx], stakingS.rewardAmount + stakingS.rewardBonus, cs.epoch+2);
+                } else {
+                    StakingLibrary.lockRewardFor(ps.validators[idx], stakingS.rewardAmount, cs.epoch+2);
+                }
             }
         }
 
+        // Check if sender is the elected validator allowed to make the snapshot
+        require(
+            mayValidatorSnapshot(
+                int(ps.ethdkgDiamond.numberOfRegistrations()),
+                int(ps.ethdkgDiamond.validatorIndex(msg.sender)),
+                int(ethBlocksSinceLastSnapshot) - int(ss.snapshotDesperationDelay),
+                keccak256(_bclaims),
+                int(ss.snapshotDesperationFactor)
+            ),
+            "validator not among allowed"
+        );
+
+        // determine whether validators changed and thus ethDKG needs to be rerun
         bool reinitEthdkg;
         if (ss.validatorsChanged) {
             reinitEthdkg = true;
         }
         ss.validatorsChanged = false;
 
-        emit SnapshotTaken(chainId, cs.epoch, height, msg.sender, ss.validatorsChanged);
+        emit SnapshotTaken(chainId, cs.epoch, madHeight, msg.sender, ss.validatorsChanged);
 
         cs.epoch++;
 
         return reinitEthdkg;
     }
 
+    function mayValidatorSnapshot(int numValidators, int myIdx, int blocksSinceDesperation, bytes32 blsig, int desperationFactor) internal pure returns (bool) {        
+        int numValidatorsAllowed = 1;
+        
+        for (int i = blocksSinceDesperation; i >= 0;) {
```

Comment:
What is exactly `i` in this context and what exactly this for loop is doing? I mean, I can see that this for loop is computing some sort of mathematical equation. A suggestion, solidity has `while` loops (different from golang), maybe rewriting this loop as while loop with some renames, can make this functionality clearer to outside readers.